### PR TITLE
Add `signal` option and remove `abort()` method

### LINKED ISSUE #51: Support `AbortController`
https://developer.mozilla.org/en-US/docs/Web/API/AbortController

And deprecate the `.abort()` method.

### PR DESCRIPTION
**Pull request by liuhanqu**
Fixes #51

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -91,15 +84,16 @@ export default function pThrottle({limit, interval, strict, onDelay}) {
 			});
 		};
 
-		throttled.abort = () => {
+		signal?.throwIfAborted();
+		signal?.addEventListener('abort', () => {
```
Comment (sindresorhus) [inline index.js:88]: You need to remove this listener when the function is done, so not to create event listener leaks.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/p-throttle/pull/53#discussion_r1613822091 is not done

File: index.js [hunk 2]
```diff
@@ -91,15 +84,18 @@ export default function pThrottle({limit, interval, strict, onDelay}) {
 			});
 		};
 
-		throttled.abort = () => {
-			for (const timeout of queue.keys()) {
-				clearTimeout(timeout);
-				queue.get(timeout)(new AbortError());
-			}
+		signal?.throwIfAborted();
+		if (signal && !signal.aborted) {
```
Comment (sindresorhus) [inline index.js:88]: The aborted check is moot as `throwIfAborted` already takes care of that.

File: index.js (hunk 1 above)
Comment (liuhanqu) [reply index.js]: It seem to have no way to know that the function is done?

**Comment by sindresorhus** [comment]
I think you could use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry to unsubscribe when the `throttle` function is garbage-collected.

File: index.js [hunk 3]
```diff
@@ -91,16 +84,26 @@ export default function pThrottle({limit, interval, strict, onDelay}) {
...
 		};
 
+		if (signal) {
+			signal.throwIfAborted();
+			signal.addEventListener('abort', aborted, {once: true});
+
+			const registry = new FinalizationRegistry(() => {
+				signal.removeEventListener('abort', aborted);
+			});
+			registry.register(signal, 'signal');
```
Comment (sindresorhus) [inline index.js:104]: This is incorrect. It's the `throttled` function that needs to be the target.

And the registry should be defined at the top level:

```js
const registry = new FinalizationRegistry(({signal, aborted}) => {
    signal.removeEventListener('abort', aborted);
});
```

File: index.js [hunk 4]
```diff
@@ -67,6 +60,10 @@ export default function pThrottle({limit, interval, strict, onDelay}) {
 
 	const getDelay = strict ? strictDelay : windowedDelay;
 
+	const registry = new FinalizationRegistry(({signal, aborted}) => {
+		signal?.removeEventListener('abort', aborted);
+	});
```
Comment (sindresorhus) [inline index.js:65]: This is not the top-level.

**Comment by sindresorhus** [comment]
You need to fix the merge conflict.

File: index.d.ts [hunk 5]
```diff
@@ -43,6 +32,36 @@ export type Options = {
...
+		console.log('Executing...');
+	});
+
+	await throttled();
+	await throttled();
+	controller.abort('aborted')
+	await throttled();
+	//=> Executing...
+	//=> Executing...
+	//=> Promise rejected with reason `aborted`
```
Comment (sindresorhus) [inline index.d.ts:60]: This example is missing from the readme.

**Pull request merged**